File: wrappers/kotlin/orm/src/main/kotlin/life/sekejap/Query.kt
package life.sekejap

import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * A composable, typed, multi-model query over one collection producing `List<T>`.
 * `C` is the generated columns type (the `it` in `where { it.field eq x }`). A
 * query lowers to a single SGQL statement and maps result payloads back to `T`.
 */
class Query<T, C>(
    private val db: Sekejap,
    private val collection: String,
    private val columns: C,
    private val fromJson: (JsonObject) -> T,
) {
    private var where: Filter? = null
    private val extraWhere = mutableListOf<String>()
    private var orderBy: String? = null
    private var desc = false
    private val rankTerms = mutableListOf<String>()
    private var limit: Int? = null
    private var offset: Int? = null

    fun where(build: (C) -> Filter): Query<T, C> {
        val f = build(columns)
        where = where?.let { AndFilter(it, f) } ?: f
        return this
    }

    fun sortBy(desc: Boolean = false, select: (C) -> Col<*>): Query<T, C> {
        orderBy = select(columns).name; this.desc = desc; return this
    }

    fun near(select: (C) -> Col<*>, point: GeoPoint, metres: Double): Query<T, C> {
        extraWhere.add("ST_DWithin(${select(columns).name}, ${point.toSql()}, $metres)")
        return this
    }

    fun matchText(select: (C) -> Col<*>, terms: String): Query<T, C> {
        extraWhere.add("BM25(${select(columns).name}, ${str(terms)}) > 0.0")
        return this
    }

    fun rankByText(select: (C) -> Col<*>, terms: String, weight: Double = 1.0, normalized: Boolean = true): Query<T, C> {
        val fn = if (normalized) "BM25_NORM" else "BM25"
        rankTerms.add("$fn(${select(columns).name}, ${str(terms)}) * $weight")
        return this
    }

    fun rankByVector(select: (C) -> Col<*>, vector: List<Double>, metric: VectorMetric = VectorMetric.COSINE, weight: Double = 1.0): Query<T, C> {
        rankTerms.add("${metric.fn}(${select(columns).name}, ${vec(vector)}) * $weight")
        return this
    }

    fun limitTo(n: Int): Query<T, C> { limit = n; return this }
    fun offsetBy(n: Int): Query<T, C> { offset = n; return this }

    // ── lowering ──────────────────────────────────────────────────────────────

    /** Flatten top-level ANDs so `value >= x AND value <= y` stays a flat
     *  conjunction (no wrapping parens) — the form the btree range scan detects. */
    private fun clauses(ctx: SqlContext): List<String> {
        val out = mutableListOf<String>()
        fun flatten(f: Filter) {
            if (f is AndFilter) { flatten(f.a); flatten(f.b) } else out.add(f.render(ctx))
        }
        where?.let { flatten(it) }
        out.addAll(extraWhere)
        return out
    }

    private fun build(projection: String = "*"): Pair<String, String> {
        val ctx = SqlContext()
        val sb = StringBuilder("SELECT $projection FROM $collection")
        val cs = clauses(ctx)
        if (cs.isNotEmpty()) sb.append(" WHERE ").append(cs.joinToString(" AND "))
        if (rankTerms.isNotEmpty()) sb.append(" ORDER BY ").append(rankTerms.joinToString(" + ")).append(" DESC")
        else orderBy?.let { sb.append(" ORDER BY ").append(it).append(if (desc) " DESC" else " ASC") }
        limit?.let { sb.append(" LIMIT ").append(it) }
        offset?.let { sb.append(" OFFSET ").append(it) }
        return sb.toString() to ctx.paramsJson()
    }

    fun find(): List<T> {
        val (sql, params) = build()
        val rows = Json.parseToJsonElement(SekejapNative.queryParams(db.handle, sql, params)) as JsonArray
        return rows.map { fromJson((it.jsonObject["payload"]!!).jsonObject) }
    }

    fun findFirst(): T? {
        val saved = limit; limit = 1
        val r = find(); limit = saved
        return r.firstOrNull()
    }

    fun count(): Int {
        val (sql, params) = build(projection = "COUNT(*) AS n")
        val rows = Json.parseToJsonElement(SekejapNative.queryParams(db.handle, sql, params)) as JsonArray
        if (rows.isEmpty()) return 0
        return rows[0].jsonObject["payload"]!!.jsonObject["n"]!!.jsonPrimitive.int
    }

    /** `UPDATE … SET … WHERE …`; returns rows changed. */
    fun update(assignments: Map<String, Any?>): Int {
        val ctx = SqlContext()
        val sets = assignments.entries.joinToString(", ") { "${it.key} = ${ctx.placeholder(it.value)}" }
        val sb = StringBuilder("UPDATE $collection SET $sets")
        val cs = clauses(ctx)
        if (cs.isNotEmpty()) sb.append(" WHERE ").append(cs.joinToString(" AND "))
        return SekejapNative.executeParams(db.handle, sb.toString(), ctx.paramsJson()).toInt()
    }

    /** `DELETE FROM … WHERE …`; returns rows deleted. */
    fun deleteAll(): Int {
        val ctx = SqlContext()
        val sb = StringBuilder("DELETE FROM $collection")
        val cs = clauses(ctx)
        if (cs.isNotEmpty()) sb.append(" WHERE ").append(cs.joinToString(" AND "))
        return SekejapNative.executeParams(db.handle, sb.toString(), ctx.paramsJson()).toInt()
    }

    /**
     * Reactive results: emits the current list immediately, then a fresh list
     * each time a committed change touches this collection. Cancelling the
     * collection releases the native listener. A dedicated daemon thread parks
     * on the native change feed (blocking JNI), so cancellation is delivered via
     * [SekejapNative.watchClose]'s stop signal, not by interrupting the thread.
     */
    fun watch(): Flow<List<T>> = callbackFlow {
        val wh = SekejapNative.watchOpen(db.handle)
        trySend(find())
        val worker = Thread {
            try {
                while (true) {
                    val evJson = SekejapNative.watchNext(wh) ?: break
                    val cols = (Json.parseToJsonElement(evJson).jsonObject["collections"]!!.jsonArray)
                        .map { it.jsonPrimitive.content }
                    if (collection in cols) trySend(find())
                }
            } catch (_: Throwable) {
                // channel closed / db gone
            } finally {
                SekejapNative.watchFree(wh) // sole owner frees, after watchNext returned null
            }
        }.apply { isDaemon = true; name = "sekejap-watch-$collection"; start() }

        awaitClose {
            // Unsubscribe + wake the parked watchNext (which lets `worker` exit).
            SekejapNative.watchClose(db.handle, wh)
        }
    }

    private fun str(s: String) = "'${s.replace("'", "''")}'"
    private fun vec(v: List<Double>) = "[${v.joinToString(", ")}]"
}

File: wrappers/dart/lib/src/orm/query.dart
/// The typed, multi-model query builder base.
///
/// Parameterised by the entity type `T` and its generated columns type `C`, so
/// every builder method — scalar `where`/`sortBy`, spatial `near`, text
/// `matchText`, and vector/text `rankBy…` — is typed against the entity's
/// columns. A query lowers to a single SGQL statement (scalar predicates as
/// `$1`, `$2` bindings; spatial/vector/text as inline literals) and maps result
/// payloads back to `T`. `.find()` runs it once; `.watch()` re-runs it whenever
/// a committed change touches the collection.
library;

import 'dart:async';
import 'dart:convert';

import '../rust/api/simple.dart';
import '../watch.dart';
import 'filter.dart';

/// A WGS84 longitude/latitude point for spatial predicates.
class GeoPoint {
  final double lon;
  final double lat;
  const GeoPoint(this.lon, this.lat);
  String toSql() => 'POINT($lon $lat)';
}

/// Vector distance/similarity used for ranking. `cosine` and `dot` are
/// similarity-like (higher is nearer, ranked DESC).
enum VectorMetric {
  cosine('VECTOR_COSINE'),
  dot('VECTOR_DOT');

  final String fn;
  const VectorMetric(this.fn);
}

/// A composable, typed, multi-model query over one collection producing
/// `List<T>`. `C` is the generated columns type (the `d` in `(d) => d.field`).
class Query<T, C> {
  final SekejapDb db;
  final String collection;
  final C columns;
  final T Function(Map<String, dynamic> payload) fromJson;

  Filter? _where;
  final List<String> _extraWhere = []; // spatial/text predicates (inline literals)
  String? _orderBy; // scalar ORDER BY field
  bool _desc = false;
  final List<String> _rankTerms = []; // weighted score terms → ORDER BY sum DESC
  int? _limit;
  int? _offset;

  Query(this.db, this.collection, this.columns, this.fromJson);

  // ── scalar ────────────────────────────────────────────────────────────────

  Query<T, C> where(Filter Function(C c) build) {
    final f = build(columns);
    _where = _where == null ? f : _where! & f;
    return this;
  }

  Query<T, C> sortBy(Col Function(C c) select, {bool desc = false}) {
    _orderBy = select(columns).name;
    _desc = desc;
    return this;
  }

  Query<T, C> limitTo(int n) {
    _limit = n;
    return this;
  }

  Query<T, C> offsetBy(int n) {
    _offset = n;
    return this;
  }

  // ── spatial ─────────────────────────────────────────────────────────────────

  /// Keep rows whose [select] geometry lies within [metres] of [point].
  Query<T, C> near(Col Function(C c) select, GeoPoint point,
      {required double metres}) {
    final col = select(columns).name;
    _extraWhere.add('ST_DWithin($col, ${point.toSql()}, $metres)');
    return this;
  }

  // ── text ────────────────────────────────────────────────────────────────────

  /// Keep rows whose [select] text field matches [terms] (BM25 score > 0).
  Query<T, C> matchText(Col Function(C c) select, String terms) {
    final col = select(columns).name;
    _extraWhere.add("BM25($col, ${_str(terms)}) > 0.0");
    return this;
  }

  /// Add a BM25 relevance term to the ranking (normalised 0–1 by default).
  Query<T, C> rankByText(Col Function(C c) select, String terms,
      {double weight = 1.0, bool normalized = true}) {
    final col = select(columns).name;
    final fn = normalized ? 'BM25_NORM' : 'BM25';
    _rankTerms.add('$fn($col, ${_str(terms)}) * $weight');
    return this;
  }

  // ── vector ──────────────────────────────────────────────────────────────────

  /// Add a vector-similarity term to the ranking.
  Query<T, C> rankByVector(Col Function(C c) select, List<double> query,
      {VectorMetric metric = VectorMetric.cosine, double weight = 1.0}) {
    final col = select(columns).name;
    _rankTerms.add('${metric.fn}($col, ${_vec(query)}) * $weight');
    return this;
  }

  // ── lowering ────────────────────────────────────────────────────────────────

  /// Top-level AND clauses, flattened so a range like `value >= x AND value <= y`
  /// stays a flat conjunction (no wrapping parentheses) — the form the engine's
  /// btree range-scan detection recognises. Nested ORs keep their parentheses.
  List<String> _whereClauses(SqlContext ctx) {
    final clauses = <String>[];
    void flatten(Filter f) {
      if (f is AndFilter) {
        flatten(f.a);
        flatten(f.b);
      } else {
        clauses.add(f.render(ctx));
      }
    }

    if (_where != null) flatten(_where!);
    clauses.addAll(_extraWhere);
    return clauses;
  }

  /// The SGQL statement and its JSON-encoded scalar parameter list.
  (String sql, String paramsJson) build({String projection = '*'}) {
    final ctx = SqlContext();
    final sb = StringBuffer('SELECT $projection FROM $collection');

    final clauses = _whereClauses(ctx);
    if (clauses.isNotEmpty) sb.write(' WHERE ${clauses.join(' AND ')}');

    if (_rankTerms.isNotEmpty) {
      sb.write(' ORDER BY ${_rankTerms.join(' + ')} DESC');
    } else if (_orderBy != null) {
      sb.write(' ORDER BY $_orderBy ${_desc ? 'DESC' : 'ASC'}');
    }
    if (_limit != null) sb.write(' LIMIT $_limit');
    if (_offset != null) sb.write(' OFFSET $_offset');
    return (sb.toString(), jsonEncode(ctx.params));
  }

  /// Apply [assignments] (`column: value`) to every matching row. Lowers to
  /// `UPDATE … SET … WHERE …`. Returns the number of rows changed.
  Future<int> update(Map<String, Object?> assignments) async {
    final ctx = SqlContext();
    final sets = [
      for (final e in assignments.entries)
        '${e.key} = ${ctx.placeholder(e.value)}'
    ].join(', ');
    final sb = StringBuffer('UPDATE $collection SET $sets');
    final clauses = _whereClauses(ctx);
    if (clauses.isNotEmpty) sb.write(' WHERE ${clauses.join(' AND ')}');
    final n = await dbExecuteParams(
        db: db, sql: sb.toString(), paramsJson: jsonEncode(ctx.params));
    return n.toInt();
  }

  /// Delete every matching row. Lowers to `DELETE FROM … WHERE …`. Returns the
  /// number of rows deleted.
  Future<int> deleteAll() async {
    final ctx = SqlContext();
    final sb = StringBuffer('DELETE FROM $collection');
    final clauses = _whereClauses(ctx);
    if (clauses.isNotEmpty) sb.write(' WHERE ${clauses.join(' AND ')}');
    final n = await dbExecuteParams(
        db: db, sql: sb.toString(), paramsJson: jsonEncode(ctx.params));
    return n.toInt();
  }

  /// Run the query once.
  Future<List<T>> find() async {
    final (sql, params) = build();
    final raw = await dbQueryParams(db: db, sql: sql, paramsJson: params);
    final rows = jsonDecode(raw) as List;
    return [
      for (final row in rows)
        fromJson((row['payload'] as Map).cast<String, dynamic>())
    ];
  }

  /// Run the query and return the first row, or null.
  Future<T?> findFirst() async {
    final saved = _limit;
    _limit = 1;
    final rows = await find();
    _limit = saved;
    return rows.isEmpty ? null : rows.first;
  }

  /// Count matching rows (scalar + spatial/text predicates).
  Future<int> count() async {
    final (sql, params) = build(projection: 'COUNT(*) AS n');
    final raw = await dbQueryParams(db: db, sql: sql, paramsJson: params);
    final rows = jsonDecode(raw) as List;
    if (rows.isEmpty) return 0;
    return ((rows.first['payload'] as Map)['n'] as num).toInt();
  }

  /// Reactive results: the current list now, then a fresh list each time a
  /// committed change touches this collection. Cancelling releases the native
  /// listener.
  Stream<List<T>> watch() {
    StreamSubscription<ChangeEvent>? sub;
    late StreamController<List<T>> controller;
    controller = StreamController<List<T>>(
      onListen: () async {
        controller.add(await find());
        sub = watchChanges(db).listen((event) async {
          if (event.collections.contains(collection) && !controller.isClosed) {
            controller.add(await find());
          }
        });
      },
      onCancel: () async => sub?.cancel(),
    );
    return controller.stream;
  }

  // ── literal encoders (inline, for the multi-model functions) ────────────────

  static String _str(String s) => "'${s.replaceAll("'", "''")}'";
  static String _vec(List<double> v) => '[${v.join(', ')}]';
}
